cpp-package/example/inference/sentiment_analysis_rnn.cpp

```diff
@@ -0,0 +1,464 @@
+
+ private:
+    void LoadModel(const std::string& model_json_file);
+    void LoadParameters(const std::string& model_parameters_file);
+    void LoadDictionary(const std::string &input_dictionary);
+    inline bool FileExists(const std::string& name) {
+        struct stat buffer;
+        return (stat(name.c_str(), &buffer) == 0);
+    }
+    float PredictSentimentForOneLine(const std::string &input_line);
+    int ConvertToIndexVector(const std::string& input,
+                      std::vector<float> *input_vector);
+    int GetIndexForOutputSymbolName(const std::string& output_symbol_name);
+    float GetIndexForWord(const std::string& word);
+    int GetClosestBucketKey(int num_words);
+    std::map<std::string, NDArray> args_map;
+    std::map<std::string, NDArray> aux_map;
+    std::map<std::string, int>  wordToIndex;
+    Symbol net;
+    std::map<int, Executor*> executor_buckets;
+    Context global_ctx = Context::cpu();
+};
+
+
+/*
+ * The constructor takes the following parameters as input:
+ * 1. model_json:  The RNN model in json formatted file.
+ * 2. model_params: File containing model parameters
+ * 3. input_dictionary: File containing the word and associated index.
+ * 4. num_words: Number of words which will be used to predict the sentiment.
+ *
+ * The constructor:
+ *  1. Loads the model and parameter files.
+ *  2. Loads the dictionary file to create index to word and word to index maps.
+ *  3. For each bucket key in the input vector of bucket keys, it invokes the SimpleBind to
+ *     create the executor. The bucket key determines the length of input data required
+ *     for that executor.
+ *  4. Creates a map of bucket key to corresponding executor.
+ *  5. The model is loaded only once. The executors share the memory for the parameters.
+ */
+Predictor::Predictor(const std::string& model_json,
+                     const std::string& model_params,
+                     const std::string& input_dictionary,
+                     const std::vector<int>& bucket_keys,
+                     bool use_gpu) {
+  if (use_gpu) {
+    global_ctx = Context::gpu();
+  }
+
+  /*
+   * Load the dictionary file that contains the word and its index.
+   * The function creates word to index and index to word map. The maps are used to create index
+   * vector for the input sentence.
+   */
+  LoadDictionary(input_dictionary);
+
+  // Load the model
+  LoadModel(model_json);
+
+  // Load the model parameters.
+  LoadParameters(model_params);
+
+  // Create the executors for each bucket key. The bucket key represents the shape of input data.
+  for (int bucket : bucket_keys) {
+    args_map["data0"] = NDArray(Shape(bucket, 1), global_ctx, false);
+    args_map["data1"] = NDArray(Shape(1), global_ctx, false);
+    Executor* executor = net.SimpleBind(global_ctx, args_map, std::map<std::string, NDArray>(),
+                                std::map<std::string, OpReqType>(), aux_map);
+    executor_buckets[bucket] = executor;
+  }
```

Comment:
This won't share the memory across executors, but will allocate excessive space from which only `1 / bucket_keys.Size()` is used at any time.

Reply:
Do you have to make sure that the master executor is created with the largest bucket size? If not, wouldn't the other executors need to allocate additional memory since the master executor only allocated enough for the smaller shape?

It looks like you are looping in whatever order bucket_keys is, and it looks like you are starting from smallest and going to largest bucket size.

(@szha -- you might know better than me. I know in the executor Reshape() call this is true. I would assume it must be true in the BindEX call that is being made here, but not 100% sure)

Reply:
The shared executors that I am creating are for the keys smaller to largest.
However, I noticed that the results do not change or example does not crash even if I reverse the order.
I had single stepped through the code, the GraphExecutor::InitDataEntryMemory() uses the shared memory from the given executor.

Reply:
@leleamol  -- the question I had is not weather it will effect the correctness or cause your program to crash. I am wondering weather the current order ends up allocating more memory than you need.

I don't know for sure, but I suspect if you allocate the largest executor first, then all the rest of the executors don't have to allocate extra memory and they just share memory from the largest executor.

However, I would assume that if the smallest executor is the "master" executor, then all other executors would have to allocate extra memory because the master executor did not allocate enough to satisfy any of the other executors.

Can you re-run your test w/ changing the order, and look closely at total gpu memory allocated?

Reply:
@stephenrawls 
I tried running it on CPU and I see an improvement in the memory consumption when the order is reveresed. That is creating the executor with largest bucket first.
I have updated the PR with that change.

Reply:
Great, thanks for confirming.

Reply:
Just looked at your fix though.

Instead of changing the order of your default array, you should probably just have a check to see what the max element of your array is, and allocate that first.

Otherwise you have a hidden assumption that the bucket sizes should be in a specific order, and it is easy for users not to realize that.

Reply:
@stephenrawls 
I agree, it will imply that the order is important.
Fixed the code in latest update.

Reply:
Awesome, it looks ready to go to me. Thanks.  👍